File: go/test/docking_event_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/launch-library2-sdk"
	"github.com/voxgig-sdk/launch-library2-sdk/core"

	vs "github.com/voxgig/struct"
)

func TestDockingEventEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.DockingEvent(nil)
		if ent == nil {
			t.Fatal("expected non-nil DockingEventEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := docking_eventBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "docking_event." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set LAUNCHLIBRARY__TEST_DOCKING_EVENT_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		dockingEventRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.docking_event", setup.data)))
		var dockingEventRef01Data map[string]any
		if len(dockingEventRef01DataRaw) > 0 {
			dockingEventRef01Data = core.ToMapAny(dockingEventRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = dockingEventRef01Data

		// LIST
		dockingEventRef01Ent := client.DockingEvent(nil)
		dockingEventRef01Match := map[string]any{}

		dockingEventRef01ListResult, err := dockingEventRef01Ent.List(dockingEventRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, dockingEventRef01ListOk := dockingEventRef01ListResult.([]any)
		if !dockingEventRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", dockingEventRef01ListResult)
		}

		// LOAD
		dockingEventRef01MatchDt0 := map[string]any{
			"id": dockingEventRef01Data["id"],
		}
		dockingEventRef01DataDt0Loaded, err := dockingEventRef01Ent.Load(dockingEventRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		dockingEventRef01DataDt0LoadResult := core.ToMapAny(dockingEventRef01DataDt0Loaded)
		if dockingEventRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if dockingEventRef01DataDt0LoadResult["id"] != dockingEventRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func docking_eventBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "docking_event", "DockingEventTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read docking_event test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse docking_event test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"docking_event01", "docking_event02", "docking_event03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("LAUNCHLIBRARY__TEST_DOCKING_EVENT_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"LAUNCHLIBRARY__TEST_DOCKING_EVENT_ENTID": idmap,
		"LAUNCHLIBRARY__TEST_LIVE":      "FALSE",
		"LAUNCHLIBRARY__TEST_EXPLAIN":   "FALSE",
		"LAUNCHLIBRARY__APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["LAUNCHLIBRARY__TEST_DOCKING_EVENT_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["LAUNCHLIBRARY__TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["LAUNCHLIBRARY__APIKEY"],
			},
			extra,
		})
		client = sdk.NewLaunchLibrary2SDK(core.ToMapAny(mergedOpts))
	}

	live := env["LAUNCHLIBRARY__TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["LAUNCHLIBRARY__TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: ts/test/entity/docking_event/DockingEventEntity.test.ts

const envlocal = __dirname + '/../../../.env.local'
require('dotenv').config({ quiet: true, path: [envlocal] })

import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'


import { LaunchLibrary2SDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveDelay,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


describe('DockingEventEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when LAUNCHLIBRARY2_TEST_LIVE=TRUE.
  afterEach(liveDelay('LAUNCHLIBRARY2_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = LaunchLibrary2SDK.test()
    const ent = testsdk.DockingEvent()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.LAUNCH_LIBRARY__TEST_LIVE
    for (const op of ['list', 'load']) {
      if (maybeSkipControl(t, 'entityOp', 'docking_event.' + op, live)) return
    }

    const setup = basicSetup()
    // The basic flow consumes synthetic IDs and field values from the
    // fixture (entity TestData.json). Those don't exist on the live API.
    // Skip live runs unless the user provided a real ENTID env override.
    if (setup.syntheticOnly) {
      t.skip('live entity test uses synthetic IDs from fixture — set LAUNCH_LIBRARY__TEST_DOCKING_EVENT_ENTID JSON to run live')
      return
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select

    let docking_event_ref01_data = Object.values(setup.data.existing.docking_event)[0] as any

    // LIST
    const docking_event_ref01_ent = client.DockingEvent()
    const docking_event_ref01_match: any = {}

    const docking_event_ref01_list = await docking_event_ref01_ent.list(docking_event_ref01_match)


    // LOAD
    const docking_event_ref01_match_dt0: any = {}
    docking_event_ref01_match_dt0.id = docking_event_ref01_data.id
    const docking_event_ref01_data_dt0 = await docking_event_ref01_ent.load(docking_event_ref01_match_dt0)
    assert(docking_event_ref01_data_dt0.id === docking_event_ref01_data.id)


  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/docking_event/DockingEventTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = LaunchLibrary2SDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['docking_event01','docking_event02','docking_event03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  // Detect whether the user provided a real ENTID JSON via env var. The
  // basic flow consumes synthetic IDs from the fixture file; without an
  // override those synthetic IDs reach the live API and 4xx. Surface this
  // to the test so it can skip rather than fail.
  const idmapEnvVal = process.env['LAUNCH_LIBRARY__TEST_DOCKING_EVENT_ENTID']
  const idmapOverridden = null != idmapEnvVal && idmapEnvVal.trim().startsWith('{')

  const env = envOverride({
    'LAUNCH_LIBRARY__TEST_DOCKING_EVENT_ENTID': idmap,
    'LAUNCH_LIBRARY__TEST_LIVE': 'FALSE',
    'LAUNCH_LIBRARY__TEST_EXPLAIN': 'FALSE',
    'LAUNCH_LIBRARY__APIKEY': 'NONE',
  })

  idmap = env['LAUNCH_LIBRARY__TEST_DOCKING_EVENT_ENTID']

  const live = 'TRUE' === env.LAUNCH_LIBRARY__TEST_LIVE

  if (live) {
    client = new LaunchLibrary2SDK(merge([
      {
        apikey: env.LAUNCH_LIBRARY__APIKEY,
      },
      extra
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.LAUNCH_LIBRARY__TEST_EXPLAIN,
    live,
    syntheticOnly: live && !idmapOverridden,
    now: Date.now(),
  }

  return setup
}
  
